mhs/inbound/inbound/request/handler.py

```diff
@@ -53,7 +77,48 @@ def _send_ack(self, parsed_message: ebxml_envelope.EbxmlEnvelope):
         }
 
         ack_message = ebxml_ack_envelope.EbxmlAckEnvelope(ack_context)
-        _, _, serialized_message = ack_message.serialize()
-
-        self.set_header("Content-Type", "text/xml")
+        message_id, http_headers, serialized_message = ack_message.serialize()
+        print(http_headers)
+        for k, v in http_headers.items():
+            self.set_header(k, v)
+            
         self.write(serialized_message)
+
+    def _extract_correlation_id(self, message):
+        correlation_id = message.message_dictionary[ebxml_envelope.CONVERSATION_ID]
+        if not correlation_id:
+            correlation_id = message_utilities.MessageUtilities.get_uuid()
+            log.correlation_id.set(correlation_id)
+            logger.info('006', "Didn't receive conversation id from inbound request , so have generated a new one.")
+        else:
+            log.correlation_id.set(correlation_id)
+            logger.info('007', 'Found correlation id on inbound request.')
+
+    def _extract_message_id(self, message):
+        """
+        Extracts the message id of the inbound message, this is to be logg
+        :param message:
+        :return:
+        """
+        message_id = message.message_dictionary[ebxml_envelope.MESSAGE_ID]
+        if not message_id:
+            logger.info('008', "Didn't receive message id in inbound message")
+        else:
+            log.message_id.set(message_id)
+            logger.info('009', 'Found inbound message id on request.')
+        return message_id
+
+    def extract_ref_message(self, message):
+        """
+        Extracts the reference-to message id and assigns it as the message Id in logging
+        :param message:
+        :return: the message id the inbound message is a response to
+        """
+        try:
+            message_id = message.message_dictionary[ebxml_envelope.RECEIVED_MESSAGE_ID]
+            log.message_id.set(message_id)
+            logger.info('0008', 'Found message id on inbound message.')
+            return message_id
+        except KeyError:
+            logger.info('0007', "No Message reference found ")
+            return None
```

Comment:
At first I thought the work description would handle this but I've updated it to do a null check on the result of `extract_ref_message` and log appropriately.

Reply:
Actually this no longer applies, whilst adding tests I've found the message parser will throw an ebxml parsing exception if it can't find the particular tags, so I've been able to move the exception handling to the message parsing instead. And any `_extract_` methods are safe to assume that the value will be in that dictionary otherwise an error would have occurred earlier